Error about incompatible subclass of @final class

**Bug Report**

`mypy` reports an error because of a concern about potentially incompatible subclasses, even though the class in question is marked `@final`.

**To Reproduce**

https://mypy-play.net/?mypy=latest&python=3.13&gist=c475ec59941e76a62c4d458819ebc828

```python
from __future__ import annotations
from typing import Self, final

class A:
    pass

@final
class B(A):
    @final
    def __init__(self) -> None:
        pass

    @classmethod
    def __new__(cls: type[B]) -> B:
        """ cls is type[B] and return is B, because class is final """
        new_instance: B = super().__new__(cls)
        new_instance.__init__()
        return new_instance

@final
class B2(A):
    @final
    def __init__(self) -> None:
        pass

    @classmethod
    def __new__(cls) -> Self:
        new_instance: Self = super().__new__(cls)
        new_instance.__init__()
        return new_instance
```

**Expected Behavior**

No errors reported.

Note that I think marking `__init__` as `@final` is redundant, but I included it just to show that it doesn't make a difference. I would expect no errors if the class is marked `@final` OR if `__init__` is marked `@final`. Or both, of course.

I also think the first case, where I explicitly annotate the `B` types, should not be necessary. Because the class is `@final`, I would expect the second case using `Self` to be fine. Again, I included that just to show that the error is reported either way.

**Actual Behavior**

```text
main.py:17: error: Accessing "__init__" on an instance is unsound, since instance.__init__ could be from an incompatible subclass  [misc]
main.py:29: error: Accessing "__init__" on an instance is unsound, since instance.__init__ could be from an incompatible subclass  [misc]
```

**Your Environment**

- Mypy version used: 1.15.0
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.13

--- a/mypy/checkmember.py
+++ b/mypy/checkmember.py
@@ -308,17 +308,20 @@ def report_missing_attribute(
 def analyze_instance_member_access(
     name: str, typ: Instance, mx: MemberContext, override_info: TypeInfo | None
 ) -> Type:
-    if name == "__init__" and not mx.is_super:
-        # Accessing __init__ in statically typed code would compromise
-        # type safety unless used via super().
-        mx.fail(message_registry.CANNOT_ACCESS_INIT)
-        return AnyType(TypeOfAny.from_error)
-
-    # The base object has an instance type.
-
     info = typ.type
     if override_info:
         info = override_info
+
+    method = info.get_method(name)
+
+    if name == "__init__" and not mx.is_super and not info.is_final:
+        if not method or not method.is_final:
+            # Accessing __init__ in statically typed code would compromise
+            # type safety unless used via super() or the method/class is final.
+            mx.fail(message_registry.CANNOT_ACCESS_INIT)
+            return AnyType(TypeOfAny.from_error)
+
+    # The base object has an instance type.
 
     if (
         state.find_occurrences
@@ -329,7 +332,6 @@ def analyze_instance_member_access(
         mx.msg.note("Occurrence of '{}.{}'".format(*state.find_occurrences), mx.context)
 
     # Look up the member. First look up the method dictionary.
-    method = info.get_method(name)
     if method and not isinstance(method, Decorator):
         if mx.is_super and not mx.suppress_errors:
             validate_super_call(method, mx)

--- a/test-data/unit/check-final.test
+++ b/test-data/unit/check-final.test
@@ -1229,3 +1229,24 @@ reveal_type(B() and 42)  # N: Revealed type is "Literal[42]?"
 reveal_type(C() and 42)  # N: Revealed type is "__main__.C"
 
 [builtins fixtures/bool.pyi]
+
+[case testCanAccessFinalClassInit]
+from typing import final
+
+@final
+class FinalClass:
+    pass
+
+def check_final_class() -> None:
+    new_instance = FinalClass()
+    new_instance.__init__()
+
+class FinalInit:
+    @final
+    def __init__(self) -> None:
+        pass
+
+def check_final_init() -> None:
+    new_instance = FinalInit()
+    new_instance.__init__()
+[builtins fixtures/tuple.pyi]

